### PR TITLE
Added GraalVM configuration to create a native binary

### PR DESCRIPTION
**Pull request by michaelkebe**
Added configuration to run signal-cli as a native binary. The generated binary is about 60MB, but it doesn't need an installed JRE. Could be helpful for low performance hardware like a raspberry pi or NAS systems. Startup times are much faster then.

Couldn't test dbus integration, because I don't have a working setup for this.

### CODE REVIEW

File: graalvm-config-dir/reflect-config.json [hunk 1]
```diff
@@ -0,0 +1,1447 @@
+[
```
Comment (AsamK) [inline graalvm-config-dir/reflect-config.json:1]: this file is quite large, is it autogenerated?

File: build.gradle [hunk 2]
```diff
@@ -1,6 +1,9 @@
-apply plugin: 'java'
-apply plugin: 'application'
-apply plugin: 'eclipse'
+plugins {
+    id 'java'
+    id 'application'
+    id 'eclipse'
+    id 'org.mikeneck.graalvm-native-image' version '1.1.0'
```
Comment (AsamK) [inline build.gradle:5]: I'd prefer to not add an additional plugin to the `build.gradle`.
Is it possible to just use a shell script that calls `gradle build` and `native-image`?
Or what's the advantage of the gradle plugin?

Comment (michaelkebe) [reply build.gradle:5]: Yes, I think this is possible. See

https://www.graalvm.org/reference-manual/native-image/BuildConfiguration/#building-native-image-with-java-reflection-example

I will try to create a shell script doing exactly that.

I don't know gradle well enough. Maybe we could use a second `build.gradle` renamed to for example `graalvm-build.gradle`?

**Comment by michaelkebe** [comment]
If I run

    ./gradlew -i nativeImage

I can see the command the `graalvm-native-image` plugin is generating:


```
> Task :nativeImage
jar-file: file collection/build: org.gradle.api.internal.tasks.DefaultTaskDependency$TaskDependencySet@19a1b
Caching disabled for task ':nativeImage' because:
  Build cache is disabled
Task ':nativeImage' is not up-to-date because:
  Task has failed previously.
create output directory if not exists: /root/signal-cli/build/native-image
run native-image binary.
Starting process 'command '/root/graalvm-ce-java11-21.0.0/bin/native-image''. Working directory: /root/signal-cli Command: /root/graalvm-ce-java11-21.0.0/bin/native-image -cp /root/.gradle/caches/modules-2/files-2.1/com.github.turasa/signal-service-java/2.15.3_unofficial_16/7d1347b0fd92cd48249d1f49262ea21dd95e5969/signal-service-java-2.15.3_unofficial_16.jar:/root/.gradle/caches/modules-2/files-2.1/org.bouncycastle/bcprov-jdk15on/1.68/46a080368d38b428d237a59458f9bc915222894d/bcprov-jdk15on-1.68.jar:/root/.gradle/caches/modules-2/files-2.1/net.sourceforge.argparse4j/argparse4j/0.8.1/2c8241f84acf6c924bd75be0dbd68e8d74fbcd70/argparse4j-0.8.1.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.hypfvieh/dbus-java/3.2.4/51c4b6a34c593d008661bc8ec091ef84668c607e/dbus-java-3.2.4.jar:/root/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-simple/1.7.30/e606eac955f55ecf1d8edcccba04eb8ac98088dd/slf4j-simple-1.7.30.jar:/root/.gradle/caches/modules-2/files-2.1/com.googlecode.libphonenumber/libphonenumber/8.12.6/ade471e53eb8c848f91dba4fdb2f462f8319220e/libphonenumber-8.12.6.jar:/root/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-databind/2.9.9.2/d18b4dec691df396916ecd5bd5aab99d0abdcb15/jackson-databind-2.9.9.2.jar:/root/.gradle/caches/modules-2/files-2.1/org.signal/signal-metadata-java/0.1.2/b4e0fa278a1eafa23f6ad8054a766a5d2a86c276/signal-metadata-java-0.1.2.jar:/root/.gradle/caches/modules-2/files-2.1/com.squareup.okhttp3/okhttp/4.6.0/1885ec944e0a74e78359756e88ce656fd5a596d5/okhttp-4.6.0.jar:/root/.gradle/caches/modules-2/files-2.1/org.signal/zkgroup-java/0.7.0/4399d0bf268067109c653a94d5a020eb7490eb1d/zkgroup-java-0.7.0.jar:/root/.gradle/caches/modules-2/files-2.1/org.whispersystems/signal-protocol-java/2.8.1/5055ef36539577e55c1feadf91bb1c89a7f38cd8/signal-protocol-java-2.8.1.jar:/root/.gradle/caches/modules-2/files-2.1/com.google.protobuf/protobuf-javalite/3.10.0/a7e0350493910f45a208732992f8e519ade2fde5/protobuf-javalite-3.10.0.jar:/root/.gradle/caches/modules-2/files-2.1/org.threeten/threetenbp/1.3.6/89dcc04a7e028c3c963413a71f950703cf51f057/threetenbp-1.3.6.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.hypfvieh/java-utils/1.0.6/a4328f3af85d2913d5c374b2b2b4445501d5c9b/java-utils-1.0.6.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jnr-unixsocket/0.33/3ca9b89af7f20e56b1d82e6e0e63447a1a97773b/jnr-unixsocket-0.33.jar:/root/.gradle/caches/modules-2/files-2.1/org.slf4j/slf4j-api/1.7.30/b5a4b6d16ab13e34a88fae84c35cd5d68cac922c/slf4j-api-1.7.30.jar:/root/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-annotations/2.9.0/7c10d545325e3a6e72e06381afe469fd40eb701/jackson-annotations-2.9.0.jar:/root/.gradle/caches/modules-2/files-2.1/com.fasterxml.jackson.core/jackson-core/2.9.9/bfff5af9fb8347d26bbb7959cb9b4fe9a2b0ca5e/jackson-core-2.9.9.jar:/root/.gradle/caches/modules-2/files-2.1/com.squareup.okio/okio/2.6.0/f06923d428f3c8e6f571043ec29a45d0cd9d2bf/okio-jvm-2.6.0.jar:/root/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib/1.3.71/898273189ad22779da6bed88ded39b14cb5fd432/kotlin-stdlib-1.3.71.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jnr-enxio/0.28/ff5f57e7cf37213df3aabd1c8930e737e8347181/jnr-enxio-0.28.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jnr-posix/3.0.58/86c94572a7bf74201576ceadcd29fb0bc50ee056/jnr-posix-3.0.58.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jnr-ffi/2.1.15/d055065db6e695e5c17001f8f95e6620a76eeb77/jnr-ffi-2.1.15.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jnr-constants/0.9.15/755a72c18be533af46337776eb4285b1aa45f229/jnr-constants-0.9.15.jar:/root/.gradle/caches/modules-2/files-2.1/org.whispersystems/curve25519-java/0.5.0/787dab46ed8e6f8adf9bbbb945145d503b7f181a/curve25519-java-0.5.0.jar:/root/.gradle/caches/modules-2/files-2.1/org.jetbrains.kotlin/kotlin-stdlib-common/1.3.71/e71c3fef58e26affeb03d675e91fd8abdd44aa7b/kotlin-stdlib-common-1.3.71.jar:/root/.gradle/caches/modules-2/files-2.1/org.jetbrains/annotations/13.0/919f0dfe192fb4e063e7dacadee7f8bb9a2672a9/annotations-13.0.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jffi/1.2.23/59d5e3e86e2583fba0cf04d4f126a5205a24c4b6/jffi-1.2.23.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jffi/1.2.23/ea93fbd2137c797ed8a686737e8bdfeead20f1b1/jffi-1.2.23-native.jar:/root/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-commons/7.1/431dc677cf5c56660c1c9004870de1ed1ea7ce6c/asm-commons-7.1.jar:/root/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-util/7.1/5b0b0f8cdb6c90582302ffcf5c20447206122f48/asm-util-7.1.jar:/root/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-analysis/7.1/379e0250f7a4a42c66c5e94e14d4c4491b3c2ed3/asm-analysis-7.1.jar:/root/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm-tree/7.1/a3662cf1c1d592893ffe08727f78db35392fa302/asm-tree-7.1.jar:/root/.gradle/caches/modules-2/files-2.1/org.ow2.asm/asm/7.1/fa29aa438674ff19d5e1386d2c3527a0267f291e/asm-7.1.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jnr-a64asm/1.0.0/a1cb8dbe71b5a6a0288043c3ba3ca64545be165/jnr-a64asm-1.0.0.jar:/root/.gradle/caches/modules-2/files-2.1/com.github.jnr/jnr-x86asm/1.0.2/6936bbd6c5b235665d87bd450f5e13b52d4b48/jnr-x86asm-1.0.2.jar:/root/signal-cli/build/libs/signal-cli-0.7.4.jar -H:Path=/root/signal-cli/build/native-image -H:Name=signal-cli -H:JNIConfigurationFiles= -H:DynamicProxyConfigurationFiles= -H:ReflectionConfigurationFiles= -H:ResourceConfigurationFiles= --no-fallback --allow-incomplete-classpath --report-unsupported-elements-at-runtime --enable-url-protocols=http,https --enable-https --enable-all-security-services -H:ResourceConfigurationFiles=graalvm-config-dir/resource-config.json -H:ReflectionConfigurationFiles=graalvm-config-dir/reflect-config.json org.asamk.signal.Main
Successfully started process 'command '/root/graalvm-ce-java11-21.0.0/bin/native-image''
```

The output contains the argument for the `-cp` option (a long list of all the jar dependencies).

Is there a command in gradle to give me exactly this list? Then I could generate this list with gradle without the graalvm-native-image plugin and run the `native-image` command directly from a shell script.

**Comment by michaelkebe** [comment]
Found a way in build.gradle:
```
task printRuntimeClasspath {
    doLast {
        print sourceSets.main.runtimeClasspath.asPath
    }
}
```

**Comment by AsamK** [comment]
Nice work, the startup time with the native-image is much better!
I'm a bit worried about the `reflect-config.json`, it was still missing some classes when I tested it. To make it really complete every code path would need to be executed with the agent, that includes all commands and all possible messages that can be received ... and it would need to be repeated after every update to keep the file up to date.
Is there some way to make the more automated?

**Comment by michaelkebe** [comment]
Thanks!

Yes, you are right, the `reflect-config.json` bothers me, too.

**The good news:**
I think we could use the unit tests for that, if there are any. You can then run them all with the `config-merge-dir` parameter, which combines the found reflection calls of different runs. More info here:

https://www.graalvm.org/reference-manual/native-image/BuildConfiguration/#assisted-configuration-of-native-image-builds

**The bad news:**
As I said earlier: There are some parts with `allowWrite`, which I added manually to make the native binary work.
Running the native binary with the missing manual parts shows an error like this:

    Cannot set final field: some.class.Name.fieldName. Enable by specifying "allowWrite" for this field in the reflection configuration.

The error message makes it pretty obvious what to add to the `reflect-config.json`.
Even if we had a working solution to generate the `reflection-config.json` it automatically with the agent, it wouldn't be perfect at the moment.
There is an closed issue which is not fixed completely.
https://github.com/oracle/graal/issues/2994


Currently I am working on a custom gradle task `buildNativeBinary` without an additional plugin dependency. So there is no need to have an extra `native-build.sh` shell script.

**Comment by AsamK** [comment]
Thanks, I've merged it and marked it as experimental in the README.
There are unfortunately no unit tests, and testing all the libsignal stuff would be quite difficult.
Maybe it would be good to have some integration tests against the signal sandbox servers that could be run automated ...

**Pull request merged**